src/main/java/org/spongepowered/common/world/schematic/SchematicTranslator.java

```diff
@@ -0,0 +1,565 @@
+        }
+        final int dataVersion = unprocessed.getInt(Constants.Sponge.Schematic.DATA_VERSION).get();
+        // DataFixer will be able to upgrade entity and tile entity data if and only if we're running a valid server and
+        // the data version is outdated.
+        // Don't run fixers for now
+        final boolean needsFixers = dataVersion < Constants.MINECRAFT_DATA_VERSION && SchematicTranslator.VANILLA_FIXER != null;
+        // TODO - DataFix, it's complicated....
+        final DataView updatedView = unprocessed;
+
+        final @Nullable DataView metadata = updatedView.getView(Constants.Sponge.Schematic.METADATA).orElse(null);
+        if (metadata != null) {
+            final Optional<DataView> dot_data = metadata.getView(DataQuery.of("."));
+            if (dot_data.isPresent()) {
+                final DataView data = dot_data.get();
+                for (final DataQuery key : data.keys(false)) {
+                    if (!metadata.contains(key)) {
+                        metadata.set(key, data.get(key).get());
+                    }
+                }
+            }
+        }
+        if (metadata != null) {
+            final String schematicName = metadata.getString(Constants.Sponge.Schematic.NAME).orElse("unknown");
+            metadata.getStringList(Constants.Sponge.Schematic.REQUIRED_MODS).ifPresent(mods -> {
+                for (final String modId : mods) {
+                    if (!Sponge.pluginManager().plugin(modId).isPresent()) {
+                        if (SchematicTranslator.MISSING_MOD_IDS.add(modId)) {
+                            SpongeCommon.getLogger().warn(
+                                "When attempting to load the Schematic: {} there is a missing modid {} some blocks/tiles/entities may not load correctly.",
+                                schematicName, modId
+                            );
+                        }
+                    }
+                }
+            });
+        }
+
+        // TODO error handling for these optionals
+        final int width = updatedView.getShort(Constants.Sponge.Schematic.WIDTH)
+            .orElseThrow(() -> new InvalidDataException("Missing value for: " + Constants.Sponge.Schematic.WIDTH));
+        final int height = updatedView.getShort(Constants.Sponge.Schematic.HEIGHT)
+            .orElseThrow(() -> new InvalidDataException("Missing value for: " + Constants.Sponge.Schematic.HEIGHT));
+        final int length = updatedView.getShort(Constants.Sponge.Schematic.LENGTH)
+            .orElseThrow(() -> new InvalidDataException("Missing value for: " + Constants.Sponge.Schematic.LENGTH));
+        if (width <= 0
+            || height <= 0
+            || length <= 0) {
+            throw new InvalidDataException(String.format(
+                "Schematic is larger than maximum allowable size (found: (%d, %d, %d) max: (%d, %<d, %<d)",
+                width, height, length, Constants.Sponge.Schematic.MAX_SIZE
+            ));
+        }
```

Comment:
This is incorrect, the spec uses `unsigned short` here, which means the negative values should be mapped to positive integers. Depending on how `getShort` works, you may need to first extract an `int`, then validate that against the `[0, 65535]` bounds, and finally encode that as a `short`.

Reply:
Should be fixed